### ISSUE TITLE
Why exactly is FIFOScheduler.state_dict and load_state_dict needed?

### ISSUE THREAD

**Issue by mseeger** [awslabs/autogluon#398]
Hello,

we are stuggling with FIFOScheduler.state_dict failing because the line

destination['searcher'] = pickle.dumps(self.searcher)

fails. This is because our searcher is a complex object, containing several members that just cannot easily be pickled.

Could somebody tell me why this state_dict method is needed? I see this could be nice to have in order to restart an aborted experiment, but we are not attempting to do that.

Could somebody help me to understand how I can avoid state_dict being called at all? I do not see a reason why this would be needed in normal mode.

**Comment by mseeger** [comment]
OK, looking at the code, this seems to be mainly used for checkpointing, so a solution for us is to switch off checkpointing.

I find it quite risky to blindly rely on pickle.dumps and pickle.loads to just work for a complex object, like a searcher or a scheduler. So I'd recommend to switch checkpointing off by default, until it is established it actually really works.

**Comment by zhanghang1989** [comment]
Hi @mseeger 
Yes, the `state` is mainly used for checkpointing. 

For complex searcher, we may try the following solution:
1). implementing ``__getstate__`` and `` __setstate__`` to make the searcher object pickle-serializible. 
https://stackoverflow.com/questions/11801284/how-to-overwrite-the-dump-load-methods-in-the-pickle-class-customizing-picklin
2). implementing ``save_state`` and ``load_stage`` for the searcher.

Let me know your thoughts.

**Comment by mseeger** [comment]
OK, it would be "nice to have" that I can restart an experiment from a checkpoint, but given this is likely quite complex, I am just wondering whether there is anything in AutoGluon right now that depends inherently on this.

We also need to be clear what constitutes the state versus the remaining arguments. In the current code, the state does not contain everything to recreate the scheduler and searcher, but only the part that is changing over time. I think this is good if the goal is to restart an experiment terminated somewhere, because then you will call it with the same remaining arguments.

In contrast, trying to *serialize* the state for much later is a much harder problem, where you hae to store everything and even worry about the code changing. This is not something we should aim for.

**Comment by zhanghang1989** [comment]
There are not much code depends on this feature. We mainly provided this for the user who want to resume some experiments.

Yeah. I agree that the current implementation does not contain everything in searcher and scheduler. Users need to call it with the same arguments. 

If we would like to store everything, we may make another experiment management object which launch the scheduler inside and store all the possible arguments (maybe through json or yml file). Let me know your thoughts.

**Comment by mseeger** [comment]
We can close this for now.
I will try something like this. I figure I can recreate a searcher object by starting from the current searcher object, as created in __init__ of the scheduler, and some standard Python data structures which can be pickled. This would circumvent the problem of having to create the searcher object from the pickled content alone, and the latter is not needed anyway.
